File: app/src/main/java/com/example/domain/AccountabilityEngine.kt
package com.example.domain

import com.example.data.model.Routine
import java.time.LocalDate
import java.time.LocalTime

sealed class DashboardState {
    data class RestDay(
        val message: String = "Rest Day • No routines scheduled"
    ) : DashboardState()

    data class Completed(
        val completedCount: Int,
        val totalPlanned: Int,
        val message: String = "🏆 All routines completed! Day complete.",
        val emoji: String = "🏆"
    ) : DashboardState()

    data class InProgress(
        val completedCount: Int,
        val totalPlanned: Int,
        val percentage: Double,
        val remainingCount: Int,
        val message: String = "⚡ Progress underway • $remainingCount remaining",
        val emoji: String = "⚡"
    ) : DashboardState()

    data class Accountability(
        val missedCount: Int,
        val totalPlanned: Int,
        val completedCount: Int,
        val percentage: Double,
        val missedRoutines: List<Routine>,
        val emoji: String, // 😠 or 😤
        val title: String, // e.g., "😠 1 routine remaining" or "😤 2 routines were left incomplete."
        val message: String
    ) : DashboardState()
}

object AccountabilityEngine {

    /**
     * Resolves the emotional accountability or celebration state for the given date.
     * Strictly enforces the rule that 😠 or 😤 are NEVER shown for:
     * - No routines planned
     * - Intentionally skipped routines
     * - Paused routines
     * - Rescheduled routines
     * - Future dates
     * - 100% completed qualifying routines
     */
    fun resolveDashboardState(
        dailyProgress: DailyProgressResult?,
        selectedDate: LocalDate,
        currentDate: LocalDate = LocalDate.now(),
        currentTime: LocalTime = LocalTime.now()
    ): DashboardState {
        if (dailyProgress == null || dailyProgress.totalPlanned == 0) {
            return DashboardState.RestDay("Rest Day • No routines scheduled")
        }

        val activePlanned = (dailyProgress.totalPlanned - dailyProgress.skippedCount).coerceAtLeast(0)

        // All scheduled routines were intentionally skipped
        if (activePlanned == 0) {
            return DashboardState.RestDay("All routines were skipped for this day")
        }

        val completedCount = dailyProgress.completedCount
        val percentage = dailyProgress.completionPercentage ?: 0.0

        // 100% qualifying routines completed
        if (completedCount >= activePlanned) {
            return DashboardState.Completed(
                completedCount = completedCount,
                totalPlanned = activePlanned,
                message = "🏆 All routines completed! Day complete.",
                emoji = "🏆"
            )
        }

        // Future dates are upcoming, not missed
        if (selectedDate.isAfter(currentDate)) {
            val remaining = (activePlanned - completedCount).coerceAtLeast(0)
            return DashboardState.InProgress(
                completedCount = completedCount,
                totalPlanned = activePlanned,
                percentage = percentage,
                remainingCount = remaining,
                message = "Upcoming plan • $remaining routine${if (remaining > 1) "s" else ""} scheduled",
                emoji = "⚡"
            )
        }

        // Identify which planned routines remain incomplete and unskipped
        val incompleteRoutines = dailyProgress.plannedRoutines.filter { routine ->
            !dailyProgress.completedRoutineIds.contains(routine.id) &&
                    !dailyProgress.skippedRoutineIds.contains(routine.id)
        }

        // Check if viewing a past day: all incomplete routines are genuinely missed
        val isPastDay = selectedDate.isBefore(currentDate)

        // Check if viewing today: determine which incomplete routines are overdue or if day is ending
        val missedRoutines = if (isPastDay) {
            incompleteRoutines
        } else {
            // Today: a routine is missed/overdue if current time has passed its scheduled time,
            // OR if it's evening / end of day (>= 18:00)
            val isEveningOrEndDay = currentTime.hour >= 18
            incompleteRoutines.filter { routine ->
                isEveningOrEndDay || currentTime.isAfter(LocalTime.of(routine.timeHour, routine.timeMinute))
            }
        }

        // If routines were genuinely missed/overdue
        if (missedRoutines.isNotEmpty()) {
            val missedCount = missedRoutines.size
            val emoji = if (missedCount >= 2) "😤" else "😠"
            val title = if (isPastDay) {
                if (missedCount == 1) "😠 1 routine was left incomplete."
                else "😤 $missedCount routines were left incomplete."
            } else {
                if (missedCount == 1) "😠 1 routine remaining"
                else "😤 $missedCount routines remaining"
            }

            val message = if (isPastDay) {
                "You didn't finish this day's plan. Get back on track."
            } else {
                "You planned this. You didn't complete it. Get back on track."
            }

            return DashboardState.Accountability(
                missedCount = missedCount,
                totalPlanned = activePlanned,
                completedCount = completedCount,
                percentage = percentage,
                missedRoutines = missedRoutines,
                emoji = emoji,
                title = title,
                message = message
            )
        }

        // Today, remaining routines are scheduled for later and day hasn't ended yet
        val remainingCount = (activePlanned - completedCount).coerceAtLeast(0)
        return DashboardState.InProgress(
            completedCount = completedCount,
            totalPlanned = activePlanned,
            percentage = percentage,
            remainingCount = remainingCount,
            message = "⚡ Progress underway • $remainingCount remaining • Keep going!",
            emoji = "⚡"
        )
    }
}

File: src/domain/AccountabilityEngine.ts
import { DailyProgressResult, DashboardState, StreakStats, UserSettings, CelebrationEvent } from '../types';
import { ProgressCalculationEngine } from './ProgressCalculationEngine';

export class AccountabilityEngine {
  /**
   * Evaluates the current state of routines and determines the high-level dashboard accountability state.
   */
  static resolveDashboardState(
    progress: DailyProgressResult | null,
    streakStats: StreakStats,
    date: string,
    settings: UserSettings
  ): DashboardState {
    if (!progress || progress.totalPlanned === 0) {
      return {
        type: 'REST_DAY',
        title: 'Rest Day',
        message: 'No routines scheduled for this date. Rest, recover, and recharge.',
        emoji: '🌿',
      };
    }

    const isAllDone = progress.completedCount >= progress.totalPlanned;
    if (isAllDone) {
      return {
        type: 'COMPLETED',
        title: 'Day Complete!',
        message: `100% completed • Keep the streak alive! (${streakStats.currentStreak} day streak)`,
        emoji: '🏆',
        streak: streakStats.currentStreak,
      };
    }

    const todayStr = ProgressCalculationEngine.getTodayStr();
    const isPastDate = ProgressCalculationEngine.compareDates(date, todayStr) < 0;

    // Filter unfinished routines
    const unfinished = progress.plannedRoutines.filter(
      (r) => !progress.completedRoutineIds.has(r.id) && !progress.skippedRoutineIds.has(r.id)
    );

    if (isPastDate) {
      // Any past date with unfinished tasks triggers accountability
      return {
        type: 'ACCOUNTABILITY',
        title: 'Missed Routines',
        message: `${unfinished.length} routine${unfinished.length > 1 ? 's were' : ' was'} missed on this day. Reclaim your rhythm!`,
        emoji: '⚠️',
        missedRoutines: unfinished,
        severity: 'STRICT',
      };
    }

    // It's today: check current time
    const now = new Date();
    const currentHour = now.getHours();
    const currentMinute = now.getMinutes();

    // Routine is overdue if its scheduled time has passed by more than 30 minutes
    const overdueRoutines = unfinished.filter((r) => {
      const routineMinutes = r.timeHour * 60 + r.timeMinute;
      const currentMinutes = currentHour * 60 + currentMinute;
      return currentMinutes > routineMinutes + 30;
    });

    if (overdueRoutines.length > 0 || currentHour >= 21) {
      const severity = settings.strictAccountability
        ? 'STRICT'
        : overdueRoutines.length >= 3
        ? 'FIRM'
        : 'MILD';

      const feedback = this.getAccountabilityFeedback(severity, overdueRoutines.length);

      return {
        type: 'ACCOUNTABILITY',
        title: feedback.title,
        message: feedback.message,
        emoji: feedback.emoji,
        missedRoutines: overdueRoutines.length > 0 ? overdueRoutines : unfinished,
        severity,
      };
    }

    // Still early / in progress
    const remaining = progress.totalPlanned - progress.completedCount;
    return {
      type: 'IN_PROGRESS',
      title: 'In Progress',
      message: `${progress.completedCount} of ${progress.totalPlanned} completed • ${remaining} remaining today`,
      emoji: '⚡',
      remainingCount: remaining,
    };
  }

  /**
   * Generates accountability feedback messages based on strictness level.
   */
  static getAccountabilityFeedback(
    severity: 'MILD' | 'FIRM' | 'STRICT',
    overdueCount: number
  ): { title: string; message: string; emoji: string } {
    switch (severity) {
      case 'STRICT':
        return {
          title: 'Accountability Alert',
          message: `${overdueCount} routine${overdueCount > 1 ? 's are' : ' is'} overdue. Consistency defines character. Take action now.`,
          emoji: '🚨',
        };
      case 'FIRM':
        return {
          title: 'Falling Behind Schedule',
          message: `You have ${overdueCount} pending routine${overdueCount > 1 ? 's' : ''}. Regain momentum before the day ends!`,
          emoji: '⏳',
        };
      case 'MILD':
      default:
        return {
          title: 'Action Needed',
          message: `${overdueCount} routine${overdueCount > 1 ? 's are' : ' is'} past scheduled time. You can still complete them today.`,
          emoji: '🎯',
        };
    }
  }

  /**
   * Helper to check celebration events.
   */
  static checkDailyCelebration(
    progress: DailyProgressResult,
    currentStreak: number
  ): CelebrationEvent | null {
    if (progress.totalPlanned > 0 && progress.completedCount >= progress.totalPlanned) {
      return {
        type: 'DAILY_COMPLETE',
        date: progress.date,
        totalRoutines: progress.totalPlanned,
        currentStreak,
      };
    }
    return null;
  }
}
